### ISSUE TITLE
Kotlin and the future of the plugin

### ISSUE THREAD

**Issue by klazuka** [durkiewicz/elm-plugin#64]
@durkiewicz I'm working on support for Structure View, and I'm doing it in Kotlin—a language which I know much better than Java, and whose Collections library I find much simpler than Java 8 Streams.

I have 2 questions:

1. Would you reject a PR written in Kotlin?
2. Do you see yourself working on this plugin much more in the future?

**Comment by durkiewicz** [comment]
Thank you for your pull request. I don't know Kotlin but I'm not a Java expert neither. I'm not against writing the plugin in Kotlin but all the code would need to be converted to Kotlin. I don't want to have different parts of the project written in different languages. Imagine what happens when someone does a pull request with Scala code, someone else with Frege, and so on. The project as a whole will become difficult to understand and maintain. I would like to avoid that.

Therefore I will take your code and convert it to Java, if you don't mind. Unless you want to convert the whole project to Kotlin? I'll try to find time to do that in July.

Regarding the future of the plugin: I would like to eliminate the issue with multiplied references (which would unblock the functionality of finding references)

**Comment by klazuka** [comment]
You're welcome to convert the Structure View code to Java, but I'd rather that if you're going to put some time into the project, it should be to fix the nested/duplicated reference bug. I will convert the Structure View code to Java so that we can get something shipped in the near-term.

I agree that it's not desirable to having a mixed language project, especially because the interoperability between Java 8 Streams and Kotlin Collections is not so good. The Kotlin IntelliJ plugin can convert Java code to Kotlin, but it will need some tweaks. I'm going to have some (rare) free-time during the first 2 weeks of July, so I can work on Kotlin-ifying the codebase then.

**Comment by klazuka** [comment]
Yesterday I did 2 things related to this:

- Ported the StructureView integration [from Kotlin to Java](https://github.com/durkiewicz/elm-plugin/pull/65/commits/d7cf0ce653815468d08f9ffef12cf51cd40a1ebb). The idea being that we can get this merged and shipped without getting slowed down by a wider discussion about Kotlin.
- Ported ElmAddImportHelper [from Java to Kotlin](https://github.com/klazuka/elm-plugin/commit/8e437e97220706de865c017dfdf8782c704ce492). The purpose of this exercise was to get a feel for how difficult it would be and how the code would look.

Overall, I'm really happy with how `ElmAddImportHelper` turned out in Kotlin. The type inference on local variables really cuts down on a lot of noise. There's also a lot less noise when working with optionals due to Kotlin's `?` method chaining syntax and `?:` operator for defaults/throwing.

It took me less than an hour to do the work for this single file. The Kotlin plugin in IntelliJ does a pretty nice job automatically converting Java to Kotlin, but I went over it by hand to make it more idiomatic and to get rid of Java 8 `Optional` and `Stream` locally.

So what I'd like to do now is get the [StructureView PR](https://github.com/durkiewicz/elm-plugin/pull/65) merged. I will have some free-time next week, so I was thinking about using that time to convert a larger portion of the project to Kotlin.

**Comment by d3xter** [comment]
+1 for whole project in kotlin.

I've started implementing basic Elm-Project stuff with kotlin

**Comment by bgourlie** [comment]
@klazuka Do you plan on creating a kotlin conversion branch? I would be willing to help with the conversion effort.

**Comment by klazuka** [comment]
The free time that I thought I was going to have never actually happened. Not sure when I will be able to work on this.

One thing we should definitely do is write some unit tests for the PsiReference resolving code *before* doing any kind of conversion, so that we can have some confidence that it didn't break important/tricky code.

**Comment by klazuka** [comment]
@d3xter @bgourlie I ended up [re-writing the whole thing](https://github.com/klazuka/intellij-elm) in Kotlin. I decided against porting it because there were some architectural problems in the original plugin (namely the reference/resolve system).

The new plugin is available now through the JetBrains plugin repository under the name "Elm". You will have to un-install the old plugin before installing the new one.
